[LT] Compose DDP with LazyTensor

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75350

Summary:
This patch adds the support of DDP to LazyTensor. It's done in two manners:
1) Defines a LazyProcessGroup which is a subclass of dist.ProcessGroup that
overrides comm ops: allgather, allreduce, and broadcast which are needed for
DDP. What they do is basically converting any lazy tensors into cuda tensors
and then replay the comm op in nccl. This is so called fallback approach.
2) Provides an additional lazy_comm_hook for overriding the all_reduce op
in DDP. This is a tentative workaround since we can't set the Future object
in dist._Work for the first approach, which the Reducer relies on to get
the finalized gradients. To be noted, other comm ops work because the callers
just reuse the passed in tensors as the results.
3) Please refer to the test case for how to init LazyProcessGroup and register
lazy_comm_hook.

Test Plan:
python test/lazy/test_ddp.py

Differential Revision: [D35441422](https://our.internmc.facebook.com/intern/diff/D35441422)

Commit:
[LT] Compose DDP with LazyTensor

Summary:
This patch adds the support of DDP to LazyTensor. It's done in two manners:
1) Defines a LazyProcessGroup which is a subclass of dist.ProcessGroup that
overrides comm ops: allgather, allreduce, and broadcast which are needed for
DDP. What they do is basically converting any lazy tensors into cuda tensors
and then replay the comm op in nccl. This is so called fallback approach.
2) Provides an additional lazy_comm_hook for overriding the all_reduce op
in DDP. This is a tentative workaround since we can't set the Future object
in dist._Work for the first approach, which the Reducer relies on to get
the finalized gradients. To be noted, other comm ops work because the callers
just reuse the passed in tensors as the results.
3) Please refer to the test case for how to init LazyProcessGroup and register
lazy_comm_hook.

Test Plan:
LTC_TS_CUDA=1 python test/lazy/test_ddp.py
To be noted, this is tentative. Still WIP to find a way to integrate with the CI.

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35441422).

wconstab commented:
> To be noted, this is tentative. Still WIP to find a way to integrate with the CI.

which part is stopping you- the LTC_TS_CUDA env? or something else?  We can discuss options for configuring the ts backend via python bindings so cuda via py is possible instead of env.

wconstab commented on torch/_lazy/__init__.py:
```diff
@@ -20,3 +25,105 @@ def wait_device_ops(devices=None):
+    Returns:
+        A Future that contains a LazyTensor.
+    """
+    assert os.environ.get('LTC_TS_CUDA', False)
```
whattt

alanwaketan replied:
For CPU, the situation is a bit complicated. It has two related ProcessGroup, MPI and GLOO. Also I don't have any things that allow me to test CPU DDP. So, it's CUDA only now.

wconstab commented on torch/_lazy/__init__.py:
```diff
@@ -20,3 +25,105 @@ def wait_device_ops(devices=None):
+    """
+    assert os.environ.get('LTC_TS_CUDA', False)
+
+    buffer = bucket.buffer()
```
i guess buffer is just a tensor?

wconstab commented on torch/_lazy/__init__.py:
```diff
@@ -20,3 +25,105 @@ def wait_device_ops(devices=None):
+                tensor.copy_(cuda_tensor)
+        return work
+
+    def allreduce(self, tensor_list, opts=None):
```
so.. this 'allreduce' works in some cases (maybe just in test cases?) but the 'comm hook' is the main pathway that actually gets used right?  why not have the comm_hook function call this one? (work obj is an inconvenience?)

alanwaketan replied:
It works if you directly calls `dist.all_reduce(...)`. We can't let the comm_hook to call this because we can't create a work objc with and set it's Future. The comm_hook is like a tentative workaround for this.

alanwaketan commented:
> > To be noted, this is tentative. Still WIP to find a way to integrate with the CI.
> 
> which part is stopping you- the LTC_TS_CUDA env? or something else? We can discuss options for configuring the ts backend via python bindings so cuda via py is possible instead of env.

I'm thinking of utilizing the current DDP test suite. Just haven't got time to familiar myself with it.

Commit:
Update on "[LT] Compose DDP with LazyTensor"


Summary:
This patch adds the support of DDP to LazyTensor. It's done in two manners:
1) Defines a LazyProcessGroup which is a subclass of dist.ProcessGroup that
overrides comm ops: allgather, allreduce, and broadcast which are needed for
DDP. What they do is basically converting any lazy tensors into cuda tensors
and then replay the comm op in nccl. This is so called fallback approach.
2) Provides an additional lazy_comm_hook for overriding the all_reduce op
in DDP. This is a tentative workaround since we can't set the Future object
in dist._Work for the first approach, which the Reducer relies on to get
the finalized gradients. To be noted, other comm ops work because the callers
just reuse the passed in tensors as the results.
3) Please refer to the test case for how to init LazyProcessGroup and register
lazy_comm_hook.

Test Plan:
LTC_TS_CUDA=1 python test/lazy/test_ddp.py
To be noted, this is tentative. Still WIP to find a way to integrate with the CI.

Differential Revision: [D35441422](https://our.internmc.facebook.com/intern/diff/D35441422)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35441422).

Commit:
Update on "[LT] Compose DDP with LazyTensor"


Summary:
This patch adds the support of DDP to LazyTensor. It's done in two manners:
1) Defines a LazyProcessGroup which is a subclass of dist.ProcessGroup that
overrides comm ops: allgather, allreduce, and broadcast which are needed for
DDP. What they do is basically converting any lazy tensors into cuda tensors
and then replay the comm op in nccl. This is so called fallback approach.
2) Provides an additional lazy_comm_hook for overriding the all_reduce op
in DDP. This is a tentative workaround since we can't set the Future object
in dist._Work for the first approach, which the Reducer relies on to get
the finalized gradients. To be noted, other comm ops work because the callers
just reuse the passed in tensors as the results.
3) Please refer to the test case for how to init LazyProcessGroup and register
lazy_comm_hook.

Test Plan:
LTC_TS_CUDA=1 python test/lazy/test_ddp.py
To be noted, this is tentative. Still WIP to find a way to integrate with the CI.

Differential Revision: [D35441422](https://our.internmc.facebook.com/intern/diff/D35441422)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35441422).

Commit:
Update on "[LT] Compose DDP with LazyTensor"


Summary:
This patch adds the support of DDP to LazyTensor. It's done in two manners:
1) Defines a LazyProcessGroup which is a subclass of dist.ProcessGroup that
overrides comm ops: allgather, allreduce, and broadcast which are needed for
DDP. What they do is basically converting any lazy tensors into cuda tensors
and then replay the comm op in nccl. This is so called fallback approach.
2) Provides an additional lazy_comm_hook for overriding the all_reduce op
in DDP. This is a tentative workaround since we can't set the Future object
in dist._Work for the first approach, which the Reducer relies on to get
the finalized gradients. To be noted, other comm ops work because the callers
just reuse the passed in tensors as the results.
3) Please refer to the test case for how to init LazyProcessGroup and register
lazy_comm_hook.

Test Plan:
python test/lazy/test_ddp.py

Differential Revision: [D35441422](https://our.internmc.facebook.com/intern/diff/D35441422)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35441422).

Commit:
Update on "[LT] Compose DDP with LazyTensor"


Summary:
This patch adds the support of DDP to LazyTensor. It's done in two manners:
1) Defines a LazyProcessGroup which is a subclass of dist.ProcessGroup that
overrides comm ops: allgather, allreduce, and broadcast which are needed for
DDP. What they do is basically converting any lazy tensors into cuda tensors
and then replay the comm op in nccl. This is so called fallback approach.
2) Provides an additional lazy_comm_hook for overriding the all_reduce op
in DDP. This is a tentative workaround since we can't set the Future object
in dist._Work for the first approach, which the Reducer relies on to get
the finalized gradients. To be noted, other comm ops work because the callers
just reuse the passed in tensors as the results.
3) Please refer to the test case for how to init LazyProcessGroup and register
lazy_comm_hook.

Test Plan:
python test/lazy/test_ddp.py

Differential Revision: [D35441422](https://our.internmc.facebook.com/intern/diff/D35441422)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35441422).

Commit:
Update on "[LT] Compose DDP with LazyTensor"


Summary:
This patch adds the support of DDP to LazyTensor. It's done in two manners:
1) Defines a LazyProcessGroup which is a subclass of dist.ProcessGroup that
overrides comm ops: allgather, allreduce, and broadcast which are needed for
DDP. What they do is basically converting any lazy tensors into cuda tensors
and then replay the comm op in nccl. This is so called fallback approach.
2) Provides an additional lazy_comm_hook for overriding the all_reduce op
in DDP. This is a tentative workaround since we can't set the Future object
in dist._Work for the first approach, which the Reducer relies on to get
the finalized gradients. To be noted, other comm ops work because the callers
just reuse the passed in tensors as the results.
3) Please refer to the test case for how to init LazyProcessGroup and register
lazy_comm_hook.

Test Plan:
python test/lazy/test_ddp.py

Differential Revision: [D35441422](https://our.internmc.facebook.com/intern/diff/D35441422)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35441422).

Commit:
Update on "[LT] Compose DDP with LazyTensor"


Summary:
This patch adds the support of DDP to LazyTensor. It's done in two manners:
1) Defines a LazyProcessGroup which is a subclass of dist.ProcessGroup that
overrides comm ops: allgather, allreduce, and broadcast which are needed for
DDP. What they do is basically converting any lazy tensors into cuda tensors
and then replay the comm op in nccl. This is so called fallback approach.
2) Provides an additional lazy_comm_hook for overriding the all_reduce op
in DDP. This is a tentative workaround since we can't set the Future object
in dist._Work for the first approach, which the Reducer relies on to get
the finalized gradients. To be noted, other comm ops work because the callers
just reuse the passed in tensors as the results.
3) Please refer to the test case for how to init LazyProcessGroup and register
lazy_comm_hook.

Test Plan:
python test/lazy/test_ddp.py

Differential Revision: [D35441422](https://our.internmc.facebook.com/intern/diff/D35441422)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35441422).

wconstab commented on test/lazy/test_ddp.py:
```diff
@@ -0,0 +1,149 @@
+        # Somehow, LazyTensor crashes if gradient_as_bucket_view is not set.
+        # FIXME(alanwaketan): Investigate why and if we can remove this constraint.
+        model = DDP(model, gradient_as_bucket_view=True)
+        # FIXME(alanwaketan): Do we need this or not?
```
did you conclude that we do not need this?

alanwaketan replied:
This was originally added for the mark_step to avoid NaN loss issue. So I need to investigate mark_step issue to further determine.

wconstab replied:
ok, so you're investigating now? or you are planning to land first?

alanwaketan replied:
Can I land this first and then investigate as a follow up?

wconstab replied:
my general rule of thumb is that you can land with a bug/todo if you know what/why it is happening and make a good argument that it's not a big factor for users.  Hence i'd be a little concerned with 'investigating as a follow up'.  It may not be a big deal, but i'm looking to you to justify that before landing

alanwaketan replied:
The graphs produced with mark_step is vastly different than the one reproduced without mark_step. Looking into why.

wconstab commented on test/lazy/test_ddp.py:
```diff
@@ -0,0 +1,149 @@
+    loss = loss_fn(output, labels)
+    loss.backward()
+    optimizer.step()
+    # FIXME(alanwaketan): Investigate why we produce different
```
did you compare the IR dump for the two cases?

alanwaketan replied:
Not yet. I'm holding it for the follow up.

alanwaketan replied:
It looks like we might have a race condition between mark_step and allreduce.

alanwaketan replied:
If I move the mark_step after the training loop, then the problem is solved as well.

wconstab commented on test/lazy/test_ddp.py:
```diff
@@ -0,0 +1,149 @@
+    if device.type == "lazy":
+        # Somehow, LazyTensor crashes if gradient_as_bucket_view is not set.
+        # FIXME(alanwaketan): Investigate why and if we can remove this constraint.
+        model = DDP(model, gradient_as_bucket_view=True)
```
what does using 'gradient_as_bucket_view' actually mean, and can you claim that it is ok/optimal to do so for lazy?

alanwaketan replied:
According to the [documentation](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel), it should be optimal except for one case where users want to detach the gradients. 
> gradient_as_bucket_view ([bool](https://docs.python.org/3/library/functions.html#bool)) – When set to True, gradients will be views pointing to different offsets of allreduce communication buckets. This can reduce peak memory usage, where the saved memory size will be equal to the total gradients size. Moreover, it avoids the overhead of copying between gradients and allreduce communication buckets. When gradients are views, detach_() cannot be called on the gradients. If hitting such errors, please fix it by referring to the [zero_grad()](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html#torch.optim.Optimizer.zero_grad) function in torch/optim/optimizer.py as a solution. Note that gradients will be views after first iteration, so the peak memory saving should be checked after first iteration.

wconstab replied:
hmm. ok, i think it is fine to leave this way for now.

alanwaketan commented:
There is one critical difference between `mark_step` and `.cpu()`. The former will set `sync_ltc_data` which will update all aliases (https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/core/lazy_graph_executor.cpp#L781). Since DDP heavily uses views, it’s just hard to break the graph (i.e. this fallback approach). If not breaking graphs wisely, i.e., having all the aliases in the same subgraph, then the results will be incorrect. The reason why `.cpu()` works is that it never updates the aliases. (grads are updated here instead: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L1413)

So, it will be hard to make this fallback approach robust enough to handle `mark_step`. I'm closing this and then switching to work on getting c10d ops into the graph. #76722.
